### ISSUE TITLE
systemd-logind fails in fedora:31 container (regression in 243)

### ISSUE THREAD

**Issue by paravz** [systemd/systemd#13955]
**systemd version the issue has been seen with**
> systemd-243-4.gitef67743.fc31.x86_64

**Used distribution**
 >Fedora:31 container running on fedora30 using podman-1.6.2. 
> Fedora:30 container on the same fedora30 host with systemd-241 works as expected

**Expected behaviour you didn't see**
 > systemd-logind.service started and user auth operations including su, sudo work without timeout

**Unexpected behaviour you saw**
 > systemd[57]: systemd-logind.service: Failed at step NAMESPACE spawning /usr/lib/systemd/systemd-logind: Operation not supported

**Steps to reproduce the problem**
Steps are outlined in https://github.com/containers/libpod/issues/4446

**Comment by paravz** [comment]
Podman repro steps are linked above. 
Docker repro (with typical steps to start systemd containers in docker)

- Working fedora:f30

sudo docker rm -vf sysuser; sudo docker run -it -d --tmpfs /run --tmpfs /tmp -v /sys/fs/cgroup:/sys/fs/cgroup:ro --name sysuser fedora:30 /sbin/init

sudo docker exec -it sysuser bash -c "systemctl unmask systemd-logind --now; useradd user; systemctl restart systemd-logind; systemctl status -l --all systemd-logind"
```
Removed /etc/systemd/system/systemd-logind.service.
● systemd-logind.service - Login Service
   Loaded: loaded (/usr/lib/systemd/system/systemd-logind.service; static; vendor preset: disabled)
   Active: active (running) since Thu 2019-11-07 20:07:38 UTC; 10ms ago
     Docs: man:systemd-logind.service(8)
           man:logind.conf(5)
           https://www.freedesktop.org/wiki/Software/systemd/logind
           https://www.freedesktop.org/wiki/Software/systemd/multiseat
 Main PID: 55 (systemd-logind)
   Status: "Processing requests..."
    Tasks: 1 (limit: 4915)
   Memory: 1.4M
   CGroup: /system.slice/containerd.service/system.slice/systemd-logind.service
           └─55 /usr/lib/systemd/systemd-logind

Nov 07 20:07:38 8a5c6e7fa266 systemd[1]: Starting Login Service...
Nov 07 20:07:38 8a5c6e7fa266 systemd-logind[55]: New seat seat0.
Nov 07 20:07:38 8a5c6e7fa266 systemd[1]: Started Login Service.
```

- Broken fedora:31

sudo docker rm -vf sysuser; sudo docker run -it -d --tmpfs /run --tmpfs /tmp -v /sys/fs/cgroup:/sys/fs/cgroup:ro --name sysuser fedora:31 /sbin/init

sudo docker exec -it sysuser bash -c "systemctl unmask systemd-logind --now; useradd user; systemctl restart systemd-logind; systemctl status -l --all systemd-logind"

```
Removed /etc/systemd/system/systemd-logind.service.
Job for systemd-logind.service failed because the control process exited with error code.
See "systemctl status systemd-logind.service" and "journalctl -xe" for details.
● systemd-logind.service - Login Service
   Loaded: loaded (/usr/lib/systemd/system/systemd-logind.service; static; vendor preset: disabled)
   Active: activating (auto-restart) (Result: exit-code) since Thu 2019-11-07 20:04:42 UTC; 3ms ago
     Docs: man:systemd-logind.service(8)
           man:logind.conf(5)
           https://www.freedesktop.org/wiki/Software/systemd/logind
           https://www.freedesktop.org/wiki/Software/systemd/multiseat
  Process: 59 ExecStartPre=/sbin/modprobe -abq drm (code=exited, status=226/NAMESPACE)
  Process: 60 ExecStart=/usr/lib/systemd/systemd-logind (code=exited, status=226/NAMESPACE)
 Main PID: 60 (code=exited, status=226/NAMESPACE)

Nov 07 20:04:42 875aa5827ff3 systemd[1]: systemd-logind.service: Service has no hold-off time (RestartSec=0), scheduling restart.
Nov 07 20:04:42 875aa5827ff3 systemd[1]: systemd-logind.service: Scheduled restart job, restart counter is at 2.
Nov 07 20:04:42 875aa5827ff3 systemd[1]: Stopped Login Service.
```

**Comment by poettering** [comment]
does this fail in nspawn? what syscall fails precisely? can you strace this?

**Comment by paravz** [comment]
1. Strace showed these 2 failures, due to missing CAP_SYS_RESOURCE:
```
637   prctl(PR_SET_MM, PR_SET_MM_ARG_START, 0x7f5284be1000, 0, 0) = -1 EPERM (Operation not permitted)
637   prctl(PR_SET_MM, PR_SET_MM_ARG_END, 0x7f5284be1010, 0, 0) = -1 EPERM (Operation not permitted)
```

2. when running repro with this missing capability, no syscall errors in strace.
Mayb this will help:

$ sudo docker rm -vf sysuser; sudo docker run -it -d --tmpfs /run --tmpfs /tmp -v /sys/fs/cgroup:/sys/fs/cgroup:ro --name sysuser --cap-add=CAP_SYS_RESOURCE  --cap-add=CAP_SYS_PTRACE fedora:31 /sbin/init
$ sudo docker exec -it sysuser bash -c "systemctl unmask systemd-logind --now; useradd user; systemctl restart systemd-logind; systemctl status -l --all systemd-logind"

```
$ sudo docker exec -it sysuser bash -c "SYSTEMD_LOG_LEVEL=debug systemctl start systemd-logind"
Bus n/a: changing state UNSET → OPENING
Bus n/a: changing state OPENING → AUTHENTICATING
Successfully forked off '(sd-askpwagent)' as PID 88.
Executing dbus call org.freedesktop.systemd1.Manager StartUnit(systemd-logind.service, replace)
Bus n/a: changing state AUTHENTICATING → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=StartUnit cookie=1 reply_cookie=0 signature=ss erro
r-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.systemd1 destination=n/a path=n/a interface=n/a member=n/a cookie=4 reply_cookie=1 signature=o error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=GetUnit cookie=2 reply_cookie=0 signature=s error-n
ame=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.systemd1 destination=n/a path=n/a interface=n/a member=n/a cookie=10 reply_cookie=2 signature=o error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/systemd_2dlogind_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=3 reply_cookie=0 signature=ss error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.systemd1 destination=n/a path=n/a interface=n/a member=n/a cookie=11 reply_cookie=3 signature=v error-name=n/a error-message=n/a
Adding /org/freedesktop/systemd1/job/290 to the set
Got message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dlogind_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dlogind_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=2 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobNew cookie=3 reply_cookie=0 signature=uos error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dlogind_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=5 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dlogind_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=6 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=7 reply_cookie=0 signature=uoss error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/systemd_2dlogind_2eservice interface=org.freedesktop.DBus.Properties member=Get cookie=4 reply_cookie=0 signature=ss error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.systemd1 destination=n/a path=n/a interface=n/a member=n/a cookie=12 reply_cookie=4 signature=v error-name=n/a error-message=n/a
Job for systemd-logind.service failed because the control process exited with error code.
See "systemctl status systemd-logind.service" and "journalctl -xe" for details.
Unexpected job result, assuming server side newer than us: failed
Got result failed/Input/output error for job systemd-logind.service
Bus n/a: changing state RUNNING → CLOSED

```

3. container logs: note selinux and other messages
```
$  docker logs sysuser
Mount failed for selinuxfs on /sys/fs/selinux:  Operation not permitted
Failed to set up the root directory for shared mount propagation: Operation not permitted
systemd v243-4.gitef67743.fc31 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Detected architecture x86-64.
Set hostname to <f3bc34e93214>.
Initializing machine ID from random generator.
Couldn't move remaining userspace processes, ignoring: Input/output error
ldconfig.service: unit configures an IP firewall, but the local system does not support BPF/cgroup firewalling.
(This warning is only shown for the first unit using IP firewalling.)
Failed to set devices.allow on /system.slice/containerd.service/system.slice/systemd-journald.service: Operation not permitted
```

4. journalctl, note `systemd[1]: Failed to set devices.allow on /system.slice/containerd.service/system.slice/systemd-journald.service: Operation not permitted` from pid 1, which might explain why it's not in strace (?)

```
-- Logs begin at Sat 2019-11-09 07:42:43 UTC, end at Sat 2019-11-09 07:48:28 UTC. --
Nov 09 07:42:43 f3bc34e93214 systemd-journald[17]: Journal started
Nov 09 07:42:43 f3bc34e93214 systemd-journald[17]: Runtime Journal (/run/log/journal/b742a073b5dc416a8e8f6db7c29a9b37) is 8.0M, max 585.6M, 577.6M free.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Starting Flush Journal to Persistent Storage...
Nov 09 07:42:43 f3bc34e93214 systemd-journald[17]: Runtime Journal (/run/log/journal/b742a073b5dc416a8e8f6db7c29a9b37) is 8.0M, max 585.6M, 577.6M free.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Started Flush Journal to Persistent Storage.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Starting Create Volatile Files and Directories...
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Started Create Volatile Files and Directories.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Starting Rebuild Journal Catalog...
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Starting Update UTMP about System Boot/Shutdown...
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Started Update UTMP about System Boot/Shutdown.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Started Rebuild Journal Catalog.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Starting Update is Completed...
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Started Update is Completed.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Reached target System Initialization.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Started Daily Cleanup of Temporary Directories.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Reached target Timers.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Listening on D-Bus System Message Bus Socket.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Reached target Sockets.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Reached target Basic System.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Starting Permit User Sessions...
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Starting Cleanup of Temporary Directories...
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Started Permit User Sessions.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Reached target Multi-User System.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Starting Update UTMP about System Runlevel Changes...
Nov 09 07:42:43 f3bc34e93214 systemd[1]: systemd-tmpfiles-clean.service: Succeeded.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Started Cleanup of Temporary Directories.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: systemd-update-utmp-runlevel.service: Succeeded.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Started Update UTMP about System Runlevel Changes.
Nov 09 07:42:43 f3bc34e93214 systemd[1]: Startup finished in 1month 2w 6d 22h 16min 31.792s (kernel) + 365ms (userspace) = 1month 2w 6d 22h 16min 32.157s.
Nov 09 07:42:55 f3bc34e93214 systemd[1]: Reloading.
Nov 09 07:42:55 f3bc34e93214 systemd[1]: Failed to set devices.allow on /system.slice/containerd.service/system.slice/systemd-journald.service: Operation not permitted
Nov 09 07:42:55 f3bc34e93214 systemd[1]: Starting D-Bus System Message Bus...
Nov 09 07:42:55 f3bc34e93214 systemd[1]: Started D-Bus System Message Bus.
Nov 09 07:42:55 f3bc34e93214 dbus-broker-lau[50]: Ready
Nov 09 07:42:55 f3bc34e93214 useradd[49]: new group: name=user, GID=1000
Nov 09 07:42:55 f3bc34e93214 useradd[49]: new user: name=user, UID=1000, GID=1000, home=/home/user, shell=/bin/bash
Nov 09 07:42:56 f3bc34e93214 systemd[1]: Created slice User and Session Slice.
Nov 09 07:42:56 f3bc34e93214 systemd[1]: Failed to set devices.allow on /system.slice/containerd.service/system.slice/systemd-logind.service: Operation not permitted
Nov 09 07:42:56 f3bc34e93214 systemd[1]: Starting Login Service...
Nov 09 07:42:56 f3bc34e93214 systemd[60]: systemd-logind.service: Failed to set up mount namespacing: Operation not supported
Nov 09 07:42:56 f3bc34e93214 systemd[60]: systemd-logind.service: Failed at step NAMESPACE spawning /sbin/modprobe: Operation not supported
Nov 09 07:42:56 f3bc34e93214 systemd[61]: systemd-logind.service: Failed to set up mount namespacing: Operation not supported
Nov 09 07:42:56 f3bc34e93214 systemd[61]: systemd-logind.service: Failed at step NAMESPACE spawning /usr/lib/systemd/systemd-logind: Operation not supported
Nov 09 07:42:56 f3bc34e93214 systemd[1]: systemd-logind.service: Main process exited, code=exited, status=226/NAMESPACE
Nov 09 07:42:56 f3bc34e93214 systemd[1]: systemd-logind.service: Failed with result 'exit-code'.
Nov 09 07:42:56 f3bc34e93214 systemd[1]: Failed to start Login Service.
Nov 09 07:42:56 f3bc34e93214 systemd[1]: systemd-logind.service: Service has no hold-off time (RestartSec=0), scheduling restart.
Nov 09 07:42:56 f3bc34e93214 systemd[1]: systemd-logind.service: Scheduled restart job, restart counter is at 1.
Nov 09 07:42:56 f3bc34e93214 systemd[1]: Stopped Login Service.

```

5. container with --privileged fixes this issue

**Comment by poettering** [comment]
those prctl()s don't really matter. can you provide a full strace -f of PID 1 in the container when starting logind? i.e. we want to see what the process does when forked off PID 1 but before logind is started.

**Comment by paravz** [comment]
Attached is output of 2 commands from up2date f31 system, let me know if there is a better way:
`sudo podman rm -vf f31test243; sudo strace -y -f -s1028 -o f31-systemd-start.log podman run -it --name f31test243 fedora:31 /sbin/init`

`sudo podman rm -vf f31test243; sudo strace  -y -f -s1028 -Z -o f31-systemd-start-Z.log podman run -it --name f31test243 fedora:31 /sbin/init`

[f31-systemd-start-Z.log.gz](https://github.com/systemd/systemd/files/3855234/f31-systemd-start-Z.log.gz)
[f31-systemd-start.log.gz](https://github.com/systemd/systemd/files/3855235/f31-systemd-start.log.gz)

**Comment by poettering** [comment]
uh... i need the payload of PID 1 in the container, not of the container mgr. I don't know podman, but this is just too much noise (and "logind" isn't even mentioned in it), can you ask the podman people to help you with this, otherwise there's litte we can do...

(run with strace -f -s 500, but only on PID 1 in the container, and only at the moment when logind is started. Hence maybe just install strace in your container, and boot into it, log in, then do "strace -p 1 -f -s500 -o log", and paste the resulting log file somewhere)

**Comment by paravz** [comment]
Full repro:
- Start container
  `sudo docker rm -vf sysuser; sudo docker run -it -d --tmpfs /run --tmpfs /tmp -v /sys/fs/cgroup:/sys/fs/cgroup:ro --name sysuser fedora:31 /sbin/init`

  `sudo docker exec -it sysuser bash -c "systemctl unmask systemd-logind; useradd user;"`

- Start strace on docker host, attached to PID of /sbin/init of a container
  `strace -f -s500 -p1111121 -o log`

- Start in container
  `sudo docker exec -it sysuser bash -c "systemctl start systemd-logind"`

[f31-docker-host-Z.log.gz](https://github.com/systemd/systemd/files/3867307/f31-docker-host-Z.log.gz)
[f31-docker-host.log.gz](https://github.com/systemd/systemd/files/3867308/f31-docker-host.log.gz)

**Comment by poettering** [comment]
Hmm, so this is where it fails:

```
1189160 unshare(CLONE_NEWNS)            = -1 EPERM (Operation not permitted)
```

i.e. your container manager appears to block fs namespacing. Not sure why it would do that. That said, we should generally be able to deal with that, so there's a bug in systemd that we are not.

That all said, docker's cgroup handling is seriously broken as it appears. "containerd.service" is the common unit all local containers run from, no? with that being the case i wonder how cgroup paths such as /sys/fs/cgroup/unified/system.slice/containerd.service/system.slice/systemd-logind.service/ (which ones sees plenty in the logs) are put together. This means the container's cgroup tree is placed directly inside containerd.service, which means either all containers containerd.service runs will live in the same hierarchy and fight for each other, or they only allow a single container per containerd.service? 

Also, why is docker not using CLONE_NEWCGROUP? why even leak so much of the upper structure to the container? And also, what kind of fucked container security policy is that to delegate all controllers to the container (which are known borked and unsafe to delegate) but then block CLONE_NEWNS?

Anyway, docker is really borked here, they should fix their cgroup hierarchy...

**Comment by poettering** [comment]
So, this is caused by this lockdown:

11dce8e29b2786a57cab2bfb4b1a39221cf7f2db

When CLONE_NEWNS fails, systemd tries to handle this gracefully currently, by simply operating without the sandbox for the service. Except that is when there are bind mounts in place for the unit, because in that case the sandbox would not just make stuff inaccessible but possibly rearrange stuff in the hierarchy. Now, the aforementioned commit actually makes two directories bind mounts (/run and /etc), but just on themselves which shouldn#t change much. I guess a simple fix would to not consider writable bind mounts on themselves for this purposes, and just proceed without sandbox in that case too.

(a simply workaround for you is to simply allow CLONE_NEWNS in the container)

**Comment by poettering** [comment]
Fix waiting in #14090

**Comment by paravz** [comment]
i didn't figure how to add CLONE_NEWNS separately only using docker or podman, it's available only as part of CAP_SYS_ADMIN which is too broad
